Improvements in TCP connections allocation.

fjpanag opened the pull request:
## Summary

This is an attempt to fix the issues reported in #6960:
* Only a single connection is allocated, the one actually needed.
* In `tcp_free()` the connection is actually deallocated, instead of leaving it existing forever in the list.

This is only the first step, possibly with more to come (list of pre-allocated connections, maximum connections limit etc).

For the moment, this PR handles only TCP, till we determine all the details.  
Then the fix will be copied to all other connections that use similar logic.

## Impact

Improved memory utilization in TCP.

## Testing

Only minimal testing for the moment.  
More testing is expected on actual hardware soon, I will report back.

fjpanag commented:
Previously, when `CONFIG_NET_SOLINGER` was enabled, there was some functionality to drop an old connection and reuse it.

I believe this has no meaning now, as we cannot really run out of connections (as easily before at least), and since there is no free connections lists, the connection is nowhere to be returned.

Also, as there will never be any free connections, this will trigger this code every time, causing old connections to be dropped constantly, instead of only when actually needed.

anchao commented:
@fjpanag 

>> Only a single connection is allocated, the one actually needed.

On some platforms this behavior can cause the memory fragmentation, and also if you think the memory fragmentation is acceptable, please configure the numb er of connections to 1, which will consistent with your expectations.
>> In tcp_free() the connection is actually deallocated, instead of leaving it existing forever in the list.

as mentioned in #6960, please release the connections instance if the number of connections is configured to 1, otherwise keep the current behavior, in some cases memory fragmentation is unacceptable.

fjpanag commented:
> @fjpanag
> 
> > > Only a single connection is allocated, the one actually needed.
> 
> On some platforms this behavior can cause the memory fragmentation, and also if you think the memory fragmentation is acceptable, please configure the numb er of connections to 1, which will consistent with your expectations.
> 
> > > In tcp_free() the connection is actually deallocated, instead of leaving it existing forever in the list.
> 
> as mentioned in #6960, please release the connections instance if the number of connections is configured to 1, otherwise keep the current behavior, in some cases memory fragmentation is unacceptable.

As I mentioned, my intention is to add a pool of preallocated connections, to avoid heap fragmentation. So new connections will be allocated only when the pool is exhausted.

I will include this change in this PR.

fjpanag commented:
This is my second take on improving dynamically allocated connections.

Now a pool of preallocated connections is created during start.  
These connections are preferred, reducing the danger of heap fragmentation.

If nevertheless the connections are exhausted, a new one will be allocated, allowing normal operation of the system even during temporal high load.

When these "extra" connections are no longer needed, they are properly deallocated.

---

This is not fully tested yet, I am working on this and I will comment back. In the meantime please review the proposed solution.

I am planning on also adding a maximum connections limit. This way a misbehaving system, or possibly a network attack, will not exhaust the heap (which will possibly have disastrous consequences).

fjpanag commented:
Also added the maximum connections limit.

fjpanag commented:
Everything seems fine now.

Tested on a custom server with a few dozens of connections. It seems stable, and no memory leaks were observed.

xiaoxiang781216 commented on net/tcp/tcp_conn.c:
```diff
@@ -836,10 +865,25 @@ void tcp_free(FAR struct tcp_conn_s *conn)
+#ifdef CONFIG_NET_ALLOC_CONNS
+  if ((conn->flags & TCP_PREALLOC) == 0)
+    {
+      free(conn);
```
kmm_free

fjpanag replied:
Done.

xiaoxiang781216 commented on include/nuttx/net/netconfig.h:
```diff
@@ -457,7 +457,7 @@
 
 #ifndef CONFIG_NET_TCP_CONNS
-#  ifdef CONFIG_NET_TCP
+#  if defined CONFIG_NET_TCP && !defined CONFIG_NET_ALLOC_CONNS
```
```suggestion
#  if defined(CONFIG_NET_TCP) && !defined(CONFIG_NET_ALLOC_CONNS)
```

fjpanag replied:
Done.

xiaoxiang781216 commented on net/tcp/tcp_conn.c:
```diff
@@ -586,16 +586,32 @@ FAR struct tcp_conn_s *tcp_alloc_conn(void)
+
+  for (i = 0; i < CONFIG_NET_TCP_PREALLOC_CONNS; i++)
+    {
+      conn = kmm_zalloc(sizeof(struct tcp_conn_s));
```
why not reuse g_tcp_connections?

fjpanag replied:
When using `CONFIG_NET_ALLOC_CONNS` connections are dynamically allocated in the heap. I believe it is much better to have everything in the heap, than scattering things around.

It also keeps code simpler I think, and I don't see any drawbacks this way.

xiaoxiang781216 replied:
> When using `CONFIG_NET_ALLOC_CONNS` connections are dynamically allocated in the heap. I believe it is much better to have everything in the heap, than scattering things around.

kmm_malloc has the 8bytes overhead for each allocation. If you will always allocate something initially, it's better to either define a global array or allocate in a batch.

> 
> It also keeps code simpler I think,

But not, now we have two code path to initialize conn. Even worse, we have two option to CONFIG_NET_TCP_PREALLOC_CONNS and CONFIG_NET_TCP_ALLOC_CONNS for the similar config.

> and I don't see any drawbacks this way.

xiaoxiang781216 replied:
If you really  want to allocate and free conn dynamically, it could be achieved simply by:
Set CONFIG_NET_CONNS to 1 and kmm_free conn in case CONFIG_NET_CONNS equals 1 instead put it into free list.

fjpanag replied:
The reason that I do not want to use `g_tcp_connections` is because this way we will have connections allocated both in heap and BSS,
and make memory management much more complicated (linker scripts etc). Especially in arches with multiple memory regions.

It is much better to have everything in one place. Custom memory mappings and adjustment of the settings are becoming much easier.

You are right about the heap overhead. I will change this to allocate all connections in a single `kmm_zalloc()` call.

> But not, now we have two code path to initialize conn.

This is a very simple change, all contained within the same function. It is not really complicating things.

> Even worse, we have two option to CONFIG_NET_TCP_PREALLOC_CONNS and NET_TCP_CONNS for the similar config.

I did this for clarity. But it is no problem to always use `NET_TCP_CONNS`. I will do this change soon.

> If you really want to allocate and free conn dynamically, it could be achieved simply by:
Set CONFIG_NET_CONNS to 1 and kmm_free conn in case CONFIG_NET_CONNS equals 1 instead put it into free list.

As I explained before, this is a different thing. With this approach we completely lose the benefit of having preallocated connections.

xiaoxiang781216 replied:
> The reason that I do not want to use `g_tcp_connections` is because this way we will have connections allocated both in heap and BSS, and make memory management much more complicated (linker scripts etc). Especially in arches with multiple memory regions.
> 
> It is much better to have everything in one place. Custom memory mappings and adjustment of the settings are becoming much easier.
> 

If so, you can add a new option(e.g. NET_TCP_PREALLOC_CONNS to indicate the preallocated global array, but keep NET_TCP_ALLOC_CONNS as the dynamical allocation:

1. Set NET_TCP_PREALLOC_CONNS to zero to disable the global allocation
2. Set NET_TCP_ALLOC_CONNS to 1 to free conn instead put into a free list

xiaoxiang781216 replied:
> I sincerely do not understand your proposition. I need the following:
> 
> * It is impossible to know the maximum number of connections before-hand. Thus I need dynamic connections allocation.
> * I know that a temporal high-load condition may happen. After this finishes, the connections must be deallocated to allow the system to reuse the memory. It must not keep the never-to-be-used-again memory reserved!

This can simply achieve to set CONFIG_NET_CONNS to 1, and alloc/free conn.

> * I know that at least _some_ connections are always needed, and I need them preallocated to reduce heap fragmentation, enhance performance, and reduce locking of the system (as the memory manager may do).
> 

it can achieve with new NET_TCP_PREALLOC_CONNS, you can allocate from .bss or heap in batch, both approach is fine.

> All the above sound like very reasonable and possibly common requirements for a networking application. In fact, I cannot imagine a system with unknown number of connections behave any differently.

Yes, I think all is the reasonable requirement.

> Point 2 is especially important for safety and security reasons.
> 

Point 2 is important for the device which accept the connection from client, but many IoT device only contain client side functionality and never accept the new connection. In the second case, it's safe to not release conn to heap, but put into the free list:

1. Allocate conn in batch to avoid the heap overhead
2. Avoid the memory fragment by reducing malloc/free operation
3. Improve the speed

> Your proposal cannot satisfy all 3 requirements, at least as far as I can understand it.

Please reference my suggestion here:
https://github.com/apache/nuttx/pull/7525#discussion_r1038789173
And point out which's function you want can't achieve, let improve it case by case.

fjpanag commented:
I removed the extra Kconfig and all connections are allocated in a single batch.

Also I changed init a bit, so it is simpler now, and there is no code duplication.

fjpanag commented on net/tcp/tcp_conn.c:
```diff
@@ -836,10 +863,25 @@ void tcp_free(FAR struct tcp_conn_s *conn)
+#ifdef CONFIG_NET_ALLOC_CONNS
+  if ((conn->flags & TCP_PREALLOC) == 0)
+    {
+      kmm_free(conn);
```
It doesn't matter. The preallocated connections are never free'd, and they are the only ones allocated in a batch.

The later connections, that are actually free'd, are also allocated one-by-one as needed.

xiaoxiang781216 replied:
If so, you destroy NET_TCP_CONNS meaning:
```
config NET_TCP_CONNS
	int "Number of TCP/IP connections"
	default 8
	---help---
		Maximum number of TCP/IP connections (all tasks)
		Number of TCP/IP connections (all tasks).

		This number of connections will be pre-allocated during system boot.
		If NET_ALLOC_CONNS is enabled, more connections may be allocated at
		a later time, as the system needs them. Else this will be the 
		maximum number of connections avaiable to the system at all times.
```
We need keep the code consistent with config.

xiaoxiang781216 replied:
Let's summary my suggestion:

1. NET_TCP_PREALLOC_CONNS to indicate the preallocated conn from .bss section

   - zero disable the prealloation

3. NET_TCP_ALLOC_CONNS to indicate the dynamic conn from heap:

   -  0 mean disable dynamical allocation at all
   - 1 mean allocate/free from heap every time
   - 2... mean allocate batch from heap, but not return to memory

4. Remove CONFIG_NET_ALLOC_CONNS once all conn convert to this style

fjpanag replied:
OK, we agree. (Especially on point 3!)

The only issue is that this way preallocated connections will be always in BSS.  
This means that changing this value will have the effect of *moving* the connection struct from BSS to heap (as it will be allocated if it is nevertheless needed).

This may be a bit problematic in arch'es with multiple memory regions.

It is not my favorite, but I can live with that.

---

One thing that I don't truly like is the `alloc_conn` variable as it is used now (tcp_conn.c, line 628).
I wanted to mark the connection as pre-allocated within `tcp_initialize()`, but the conn struct will be memset to 0 in line 735, so this information is lost.

Do you have any suggestions for this?

xiaoxiang781216 replied:
> OK, we agree. (Especially on point 3!)
> 
> The only issue is that this way preallocated connections will be always in BSS. This means that changing this value will have the effect of _moving_ the connection struct from BSS to heap (as it will be allocated if it is nevertheless needed).
> 
> This may be a bit problematic in arch'es with multiple memory regions.
> 
> It is not my favorite, but I can live with that.
>

We can allocate the preallocated conn from heap too, it is no much difference whether they come from .bss or heap. The more important thing is never return them to heap.

> One thing that I don't truly like is the `alloc_conn` variable as it is used now (tcp_conn.c, line 628). I wanted to mark the connection as pre-allocated within `tcp_initialize()`, but the conn struct will be memset to 0 in line 735, so this information is lost.
> 
> Do you have any suggestions for this?

One simple approach is:

1. Don't preallocate conn in startup instead
2. Put conn to free list if free list is smaller than NET_TCP_PREALLOC_CONNS
3. Call kmm_free if the free list is larger than NET_TCP_PREALLOC_CONNS

So, we don't need flag to distinguish whether conn is preallocate or not.

fjpanag replied:
> One simple approach is:
> 
> 1. Don't preallocate conn in startup instead
> 2. Put conn to free list if free list is smaller than NET_TCP_PREALLOC_CONNS
> 3. Call kmm_free if the free list is larger than NET_TCP_PREALLOC_CONNS
> 
> So, we don't need flag to distinguish whether conn is preallocate or not.

This is what I thought first, but this is not working.  
We need to preserve the exact same connections allocated during init. If not, heap fragmentation will happen. The preallocated list will serve absolutely no purpose.

xiaoxiang781216 replied:
Yes, it's a problem. So, I think the better approach is move flag mark and memset to tcp_alloc_conn.

fjpanag replied:
> Yes, it's a problem. So, I think the better approach is move flag mark and memset to tcp_alloc_conn.

I also thought of that, but now `memset()` is used for all connections (even those in the free list), every time (not only during the initial allocation).

I am not sure whether code anywhere else works under the assumption that conns are `memset()` at this point, or it is just defensive coding.

Also I am not sure whether
```g_tcp_connections[i].tcpstateflags = TCP_CLOSED;```
has any meaning in line 605, or it will be overwritten by the memset (and thus it is dead code).

fjpanag commented:
Here is another take on this issue.  
Not fully tested yet, may change a few things here and there.

@xiaoxiang781216 

>     NET_TCP_PREALLOC_CONNS to indicate the preallocated conn from .bss/data or heap:
        zero disable the prealloation

Done. For the moment BSS is used, I am not sure whether I will change this to use the heap.

> NET_TCP_ALLOC_CONNS to indicate the dynamic conn from heap:
> 0 mean disable dynamical allocation at all
> 1 mean allocate/free from heap every time
> 2... mean allocate batch from heap, but not return to heap

Done. This way it satisfies everyone.

> Add NET_TCP_MAX_CONNS to limit the concurrent opened connection, actually I think it's better to impose the limitation from psocket level instead individual protocol(e.g. TCP/UDP/ICMP).

I left this in TCP level. I believe it is much better to be able to tune each protocol independently. Especially because each protocol has a different connection structure size, and different usage. For example the heap may have enough space for 50 UDP connections, but not for 50 TCP connections.

> Finally, remove NET_ALLOC_CONNS once all conn convert to this style since we can achieve the same functionality by setting NET_xxx_ALLOC_CONNS to zero or non zero.

Yes will do, in a future PR though.

fjpanag commented:
I just had the time to complete my tests, and this seems to behave as expected.

@xiaoxiang781216 What do you think, can we merge this?

Does anyone understand what is the issue with the CI?  
Shall we just restart it?

xiaoxiang781216 commented:
> I just had the time to complete my tests, and this seems to behave as expected.
> 
> @xiaoxiang781216 What do you think, can we merge this?
> 

I will take a look in tomorrow

> Does anyone understand what is the issue with the CI? Shall we just restart it?

I am trying fix ci break here: https://github.com/apache/nuttx/pull/7842

xiaoxiang781216 commented on net/tcp/tcp_conn.c:
```diff
@@ -549,16 +549,25 @@ FAR struct tcp_conn_s *tcp_alloc_conn(void)
   if (dq_peek(&g_free_tcp_connections) == NULL)
     {
+#if CONFIG_NET_TCP_MAX_CONNS > 0
+      if (dq_count(&g_active_tcp_connections) + CONFIG_NET_TCP_ALLOC_CONNS
```
```suggestion
      if (dq_count(&g_active_tcp_connections) + CONFIG_NET_TCP_PREALLOC_CONNS
```

fjpanag replied:
No, this is correct.

The calculation adds the existing connections with those that are about to be allocated. The sum of them must not exceed the limit.

I don't see the meaning  of  `CONFIG_NET_TCP_PREALLOC_CONNS` here.

xiaoxiang781216 replied:
Yes, it's my mistake.

xiaoxiang781216 commented on net/tcp/Kconfig:
```diff
@@ -51,11 +51,45 @@ config NET_TCPURGDATA
+
+config NET_TCP_ALLOC_CONNS
+	int "Dynamic TCP/IP connections allocation"
+	default 0
```
```suggestion
	default NET_TCP_PREALLOC_CONNS
```

fjpanag replied:
@xiaoxiang781216 Why this recommendation?

The default that I used make the system behave exactly as before.  
There are some statically allocated connections, and dynamic allocation is disabled.

Your proposition alters the default behavior.

Furthermore, why the preallocated connections are related to the dyn. allocation batch size?

fjpanag replied:
@anchao I disagree.

I think it is greatly beneficial to have the ability to fine tune each protocol independently.

For example, a system may have a finite number of expected UDP connections, so dyn. alloc. is disabled for safety.  
On the other hand, TCP may be unpredictable, so more connections may be needed.

xiaoxiang781216 replied:
> @xiaoxiang781216 Why this recommendation?
> 
> The default that I used make the system behave exactly as before. There are some statically allocated connections, and dynamic allocation is disabled.
> 
> Your proposition alters the default behavior.
> 

Before this change, the dynamical allocation is enabled by default, I think.

> Furthermore, why the preallocated connections are related to the dyn. allocation batch size?

let change 0 to 8.

fjpanag replied:
I see that the default was to have dynamic connections disabled:  
https://github.com/apache/nuttx/blob/master/net/Kconfig#L338

Do you insist on enabling it now?  
Maybe setting it to 1 would be safer?

xiaoxiang781216 replied:
Ok, it's fine to keep the behavior same as before.

xiaoxiang781216 commented on net/tcp/tcp_conn.c:
```diff
@@ -699,9 +717,10 @@ FAR struct tcp_conn_s *tcp_alloc(uint8_t domain)
+#if CONFIG_NET_TCP_ALLOC_CONNS > 0
   if (conn == NULL)
     {
+      alloc_conn = 1;
```
the allocation check is wrong since the free connection got above may dynamically allocate too, but put to the free list in tcp_free.

fjpanag replied:
Done.

xiaoxiang781216 commented on net/tcp/tcp_conn.c:
```diff
@@ -836,10 +862,25 @@ void tcp_free(FAR struct tcp_conn_s *conn)
+   */
+
+#if CONFIG_NET_TCP_ALLOC_CONNS == 1
+  if ((conn->flags & TCP_PREALLOC) == 0)
```
there is a simple method to know this conn come from global variable by:
```
conn >= g_tcp_connections && conn < g_tcp_connections + CONFIG_NET_TCP_PREALLOC_CONNS
```

xiaoxiang781216 commented on include/nuttx/net/netconfig.h:
```diff
@@ -456,11 +456,11 @@
-#ifndef CONFIG_NET_TCP_CONNS
-#  ifdef CONFIG_NET_TCP
-#   define CONFIG_NET_TCP_CONNS 10
+#ifndef CONFIG_NET_TCP_PREALLOC_CONNS
```
should we simply remove line 450-466?

fjpanag replied:
I think that this is temporarily needed, till `NET_ALLOC_CONNS` is removed.  
I kept this for this purpose, with the intention to remove this soon.

xiaoxiang781216 replied:
If so, let's simplify to:
```
#ifndef CONFIG_NET_TCP_PREALLOC_CONNS
#  define CONFIG_NET_TCP_PREALLOC_CONNS 0
#endif
```

anchao commented on net/tcp/tcp.h:
```diff
@@ -104,6 +104,7 @@
 /* The TCP options flags */
 
 #define TCP_WSCALE            0x01U /* Window Scale option enabled */
+#define TCP_PREALLOC          0x02U /* TCP connection is preallocated */
```
Do not use the flag of tcp option field, if the alloc policy will be applied to other protocols,  use [(struct socket_conn_s)](https://github.com/apache/nuttx/blob/master/include/nuttx/net/net.h#L192-L226)

fjpanag replied:
Fixed. Used alternative approach instead.

xiaoxiang781216 commented:
@fjpanag  please rebase to the last master, ci break has been fixed.

fjpanag commented on net/tcp/Kconfig:
```diff
@@ -51,11 +51,45 @@ config NET_TCPURGDATA
+		connection is no longer needed, it will be returned to the
+		free connections pool, and it will never be deallocated!
+
+config NET_TCP_MAX_CONNS
```
@anchao Each protocol may have a vastly different use, so different limits may be needed.

Also each connection struct has a different size, so for a given heap size you need different limits to take advantage of it all.

anchao replied:
emm ... since in this way, too many configurations will be added to each protocol, could we only changes the number of pre-allocated connections to achieve this feature?

fjpanag commented:
> @fjpanag it's better to apply the similar change to all socket conn, not only tcp.

Yes I will apply the change in all connections.  
But first let's finalize the design using TCP.

For every connection will be a new commit.  
Do you want them in the same PR? I think it is better to have them separated.

xiaoxiang781216 commented:
@fjpanag please rebase the change to the last master. macOS break is fixed here: https://github.com/apache/nuttx/pull/7849

fjpanag commented:
@xiaoxiang781216 I getting back to this issue.

What do you think is needed to be done to be closed?

xiaoxiang781216 commented:
> @xiaoxiang781216 I getting back to this issue.
> 
> What do you think is needed to be done to be closed?

Fix the break and apply the similar change to other protocol.

fjpanag commented:
@xiaoxiang781216 I getting back to this issue.



> > @xiaoxiang781216 I getting back to this issue.
> > What do you think is needed to be done to be closed?
> 
> Fix the break and apply the similar change to other protocol.

I can't see what is the CI problem. I don't see any error.

Can I have any hints on this?

btashton commented:
> @xiaoxiang781216 I getting back to this issue.
> 
> 
> 
> > > @xiaoxiang781216 I getting back to this issue.
> > > What do you think is needed to be done to be closed?
> > 
> > Fix the break and apply the similar change to other protocol.
> 
> I can't see what is the CI problem. I don't see any error.
> 
> Can I have any hints on this?


This config needs to be normalized. 

https://github.com/apache/nuttx/actions/runs/3719713789/jobs/6308762214#step:7:2400

fjpanag commented:
Is this 100% OK?  
May I copy the changes to the other protocols?

xiaoxiang781216 commented:
> Is this 100% OK? May I copy the changes to the other protocols?

It's fine since the change focus on the conn allocation which isn't relate to protocol self.

xiaoxiang781216 commented:
@fjpanag will you finish other conn types?

fjpanag commented:
> @fjpanag will you finish other conn types?

Yes, yes. I am working on this.

fjpanag commented:
Hmm, why is this check failing?  
I think that `checkpatch.sh` picks up other commits than the two actually in this PR?

xiaoxiang781216 commented:
> Hmm, why is this check failing? I think that `checkpatch.sh` picks up other commits than the two actually in this PR?

Please rebase your patch to the last master to fix this problem.

xiaoxiang781216 commented:
@fjpanag please ping you once you finish the change.

fjpanag commented:
@xiaoxiang781216 Please have a look. I think that I have finished all protocols.

I expect to do some more testing tonight on actual hardware.

If everything is OK, all that remains is to remove the `CONFIG_NET_ALLOC_CONNS` option that now is not used anywhere (and normalize a config that uses it).

xiaoxiang781216 commented:
@fjpanag thanks, I will review it tomorrow.

xiaoxiang781216 commented on net/icmp/Kconfig:
```diff
@@ -32,9 +32,45 @@ config NET_ICMP_SOCKET
+
+config NET_ICMP_MAX_CONNS
+	int "Maximum number of ICMP connections"
+	default 0
```
add dependence too

fjpanag replied:
Done.

xiaoxiang781216 commented on net/usrsock/Kconfig:
```diff
@@ -25,15 +25,49 @@ config NET_USRSOCK
+
+config NET_USRSOCK_MAX_CONNS
+	int "Maximum number of usrsock connections"
+	default 0
```
add dependence too

fjpanag replied:
Done.

xiaoxiang781216 commented on net/bluetooth/Kconfig:
```diff
@@ -22,9 +22,45 @@ config NET_BLUETOOTH
+
+config NET_BLUETOOTH_MAX_CONNS
+	int "Maximum number of Bluetooth connections"
+	default 0
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 replied:
but no change

fjpanag replied:
Isn't this OK?  
https://github.com/apache/nuttx/pull/7525/commits/d550e68264632321f12ccae5b28a24ae6da6586e#diff-789fc80906d45426aebac87a9b963c00ea7679ebe7155dbe2d8e7015e81f6ac3R58

xiaoxiang781216 replied:
Yes.

xiaoxiang781216 commented on net/icmpv6/Kconfig:
```diff
@@ -200,9 +200,45 @@ endif # NET_ICMPv6_ROUTER
+
+config NET_ICMPv6_MAX_CONNS
+	int "Maximum number of ICMPv6 connections"
+	default 0
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/ieee802154/Kconfig:
```diff
@@ -33,9 +33,45 @@ config NET_IEEE802154_FRAMELEN
+
+config NET_IEEE802154_MAX_CONNS
+	int "Maximum number of IEEE 802.15.4 connections"
+	default 0
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/pkt/Kconfig:
```diff
@@ -22,9 +22,45 @@ config NET_PKT
+
+config NET_PKT_MAX_CONNS
+	int "Maximum number of packet connections"
+	default 0
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/netlink/Kconfig:
```diff
@@ -22,11 +22,45 @@ config NET_NETLINK
+
+config NETLINK_MAX_CONNS
+	int "Maximum number of Netlink connections"
+	default 0
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on include/nuttx/net/netconfig.h:
```diff
@@ -316,16 +316,6 @@
 
-/* The maximum amount of concurrent UDP connection, Default: 10 */
-
-#ifndef CONFIG_NET_UDP_PREALLOC_CONNS
```
move to udp patch

xiaoxiang781216 commented on include/nuttx/net/netconfig.h:
```diff
@@ -444,21 +434,6 @@
 
 /* TCP configuration options */
 
-/* The maximum number of simultaneously open TCP connections.
```
move to tcp patch

xiaoxiang781216 commented on net/devif/devif_callback.c:
```diff
@@ -43,8 +43,8 @@
 
-#ifndef CONFIG_NET_ALLOC_CONNS
-static struct devif_callback_s g_cbprealloc[CONFIG_NET_NACTIVESOCKETS];
+#if CONFIG_NET_PREALLOC_SOCKETS > 0
```
should we change NET_PREALLOC_SOCKETS  to NET_PREALLOC_CALLBACK or NET_PREALLOC_DEVIF_CALLBACK?

fjpanag replied:
I could rename this, but it is currently defined in the "Socket Support" menu in Kconfig.  
Maybe it should be moved too? But I can't decide where it would be best.

xiaoxiang781216 replied:
Move to devif/Kconfig or net/kconfig?

fjpanag replied:
Just renamed for the moment. I didn't like any other place to move to.
(devif/Kconfig does not exist and devif/Kconfig would be a bit chaotic...)

xiaoxiang781216 commented on net/can/Kconfig:
```diff
@@ -30,11 +30,45 @@ config NET_CAN_HAVE_ERRORS
+
+config CAN_MAX_CONNS
+	int "Maximum number of CAN connections"
+	default 0
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/can/can_conn.c:
```diff
@@ -101,20 +101,30 @@ void can_initialize(void)
+#endif
+
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_CAN_ALLOC_CONNS);
+
```
remove the blank line

fjpanag replied:
Done.

xiaoxiang781216 commented on net/netlink/netlink_conn.c:
```diff
@@ -131,20 +132,30 @@ void netlink_initialize(void)
+#endif
+
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_NETLINK_ALLOC_CONNS);
+
```
remove the blank line

fjpanag replied:
Done.

xiaoxiang781216 commented on net/pkt/pkt_conn.c:
```diff
@@ -106,20 +106,30 @@ void pkt_initialize(void)
+#endif
+
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_PKT_ALLOC_CONNS);
+
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/ieee802154/ieee802154_conn.c:
```diff
@@ -107,20 +107,31 @@ void ieee802154_conn_initialize(void)
+#endif
+
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_IEEE802154_ALLOC_CONNS);
+
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/bluetooth/bluetooth_conn.c:
```diff
@@ -113,20 +113,30 @@ void bluetooth_conn_initialize(void)
+#endif
+
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_BLUETOOTH_ALLOC_CONNS);
+
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/icmpv6/icmpv6_conn.c:
```diff
@@ -109,13 +110,23 @@ FAR struct icmpv6_conn_s *icmpv6_alloc(void)
+#endif
+
+          conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_ICMPv6_ALLOC_CONNS);
+
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/icmp/icmp_conn.c:
```diff
@@ -109,13 +109,23 @@ FAR struct icmp_conn_s *icmp_alloc(void)
+#endif
+
+          conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_ICMP_ALLOC_CONNS);
+
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/usrsock/usrsock_conn.c:
```diff
@@ -77,20 +78,30 @@ static dq_queue_t g_active_usrsock_connections;
+#endif
+
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_USRSOCK_ALLOC_CONNS);
+
```
ditto

fjpanag replied:
Done.

xiaoxiang781216 commented on net/udp/udp_conn.c:
```diff
@@ -472,16 +472,25 @@ FAR struct udp_conn_s *udp_alloc_conn(void)
       conn = kmm_zalloc(sizeof(struct udp_conn_s) *
-                        CONFIG_NET_UDP_CONNS);
+                        CONFIG_NET_UDP_ALLOC_CONNS);
+
```
remove the blank line

fjpanag replied:
Done.

xiaoxiang781216 commented on net/tcp/tcp_conn.c:
```diff
@@ -449,16 +449,25 @@ FAR struct tcp_conn_s *tcp_alloc_conn(void)
       conn = kmm_zalloc(sizeof(struct tcp_conn_s) *
-                        CONFIG_NET_TCP_CONNS);
+                        CONFIG_NET_TCP_ALLOC_CONNS);
+
```
remove the blank line

fjpanag replied:
Done.

xiaoxiang781216 commented:
not all comments get addressed?

fjpanag commented on net/tcp/Kconfig:
```diff
@@ -50,11 +50,45 @@ config NET_TCPURGDATA
+
+config NET_TCP_MAX_CONNS
+	int "Maximum number of TCP/IP connections"
+	default 0
```
Done.

fjpanag commented on net/udp/Kconfig:
```diff
@@ -36,11 +36,45 @@ config NET_UDP_CHECKSUMS
+
+config NET_UDP_MAX_CONNS
+	int "Maximum number of UDP connections"
+	default 0
```
Done.

fjpanag commented:
> not all comments get addressed?

Working on it... I will ping you.

fjpanag commented on include/nuttx/net/netconfig.h:
```diff
@@ -451,11 +451,11 @@
-#ifndef CONFIG_NET_TCP_CONNS
-#  ifdef CONFIG_NET_TCP
-#   define CONFIG_NET_TCP_CONNS 10
+#ifndef CONFIG_NET_TCP_PREALLOC_CONNS
```
Isn't this already addressed?

xiaoxiang781216 replied:
but it's better to move the change to the first patch, instead removing in the later patch.

fjpanag replied:
I am not sure how to do this, without braking everything.  
Moving a deletion back in history?...

Is it a big deal if we leave it as is? It nevertheless works, and when the PR is merged it won't have any significance whatsoever.

xiaoxiang781216 replied:
it's very simple with rebase command:
```
git rebase --interactive HEAD~12
change the first two from pick to edit
remove CONFIG_NET_TCP_CONNS from netconfig.h and commit the patch
git rebase --continue
remove CONFIG_NET_UDP_CONNS from netconfig.h and commit the patch
git rebase --continue
```
That's it.

fjpanag replied:
Thank you for your advice. It was easy indeed, provided that you know how to do it ;)

Just pushed the change...

fjpanag commented on include/nuttx/net/netconfig.h:
```diff
@@ -318,11 +318,11 @@
-#ifndef CONFIG_NET_UDP_CONNS
-#  ifdef CONFIG_NET_UDP
-#    define CONFIG_NET_UDP_CONNS 10
+#ifndef CONFIG_NET_UDP_PREALLOC_CONNS
```
Similarly, this has been removed at a later point.

xiaoxiang781216 replied:
but it's better to move the change here directly.

xiaoxiang781216 commented:
@fjpanag CONFIG_NET_ALLOC_SOCKETS need be removed from sim/dynconns/defconfig:
```
====================================================================================
Configuration/Tool: sim/dynconns
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Building NuttX...
  Normalize sim/dynconns
49d48
< CONFIG_NET_ALLOC_SOCKETS=1
Saving the new configuration file
```

fjpanag commented:
Users defconfigs are needed to be updated after this PR.  
Maybe this must be marked to be included in the next release notes?

xiaoxiang781216 commented:
Marked.

pkarashchenko commented on net/pkt/pkt_conn.c:
```diff
@@ -164,9 +173,22 @@ void pkt_free(FAR struct pkt_conn_s *conn)
 
   memset(conn, 0, sizeof(*conn));
```
ditto

pkarashchenko commented on net/tcp/tcp_conn.c:
```diff
@@ -876,10 +889,26 @@ void tcp_free(FAR struct tcp_conn_s *conn)
+  else
+#endif
+    {
+      dq_addlast(&conn->sconn.node, &g_free_tcp_connections);
```
Do we need `memset(conn, 0, sizeof(*conn));` before adding to the queue

fjpanag replied:
I think not. It seems that `tcp_free()` sets the contents of `conn` to specific values, for example see [this](https://github.com/apache/nuttx/blob/master/net/tcp/tcp_conn.c#L894).

If someone knows for sure how this works, I can add a `memset()` and then set the flags, or whatever else is needed.

pkarashchenko commented on net/udp/udp_conn.c:
```diff
@@ -703,9 +716,23 @@ void udp_free(FAR struct udp_conn_s *conn)
 
   memset(conn, 0, sizeof(*conn));
```
ditto

pkarashchenko commented on net/usrsock/usrsock_conn.c:
```diff
@@ -143,9 +153,22 @@ void usrsock_free(FAR struct usrsock_conn_s *conn)
   nxsem_destroy(&conn->resp.sem);
   memset(conn, 0, sizeof(*conn));
```
ditto

fjpanag commented on net/ieee802154/ieee802154_conn.c:
```diff
@@ -188,9 +198,22 @@ void ieee802154_conn_free(FAR struct ieee802154_conn_s *conn)
+  else
+#endif
+    {
+      dq_addlast(&conn->sconn.node, &g_free_ieee802154_connections);
```
Yes this is a good idea, I will provide a new PR for this.

fjpanag replied:
@pkarashchenko see #8669.
